### PR TITLE
add feature importance in python

### PR DESCRIPTION
**Pull request by wxchan**
Changes:
1. add feature importance with 2 options: split and gain
2. simplify functions (list comprehension is recommended in python)
3. format comments

Discussions:
1. if user calls train() with train_data is Dataset object and categorical_feature is not None, categorical_feature won't be used for Dataset;
2. best_iteration is not right with init_model. For example, init_model has 100 trees, train 10 more iterations and meet early stopping criterion, then if use num_iteration=gbm.best_iteration in predict, it only uses 10 trees (should use 110 trees).

### CODE REVIEW

**Comment by guolinke** [comment]
1. Should we expose ```Dataset``` to the user? Or we can use another class that warp ```Dataset``` and other filed, and expose this "high level" data-set to user?
2. use Booster.predictor to fix this.

**Comment by wxchan** [comment]
Before we have better solutions, we can use examples or wiki to recommend user to use tuple. People can still use Dataset if he knows what he is doing. 

I tried to solve 2, but it's not working. The logic seems not easy.

**Comment by guolinke** [comment]
@wxchan 
1. Actually, current interfaces in engine.py and sklearn.py cannot save memory after convert to ```Dataset```. Since the reference of raw data is not freed. 
My suggestions:
    - use high level Dataset that can set X, y , categorical_features, init_predictor, other_fields. 
    - engine.py only can accept this Dataset as input. 
    - Construct real Dataset before construct Booster(in training or cv). After construct real Dataset, free raw data
    - If user want to predict Dataset without raw data, raise an exception

2. you can update the code, I can help to take a look.

**Comment by guolinke** [comment]
@wxchan OK, i will merge this first, then create another PR to refine ```Dataset```.

File: python-package/lightgbm/sklearn.py [hunk 1]
```diff
@@ -1,10 +1,10 @@
 # coding: utf-8
-# pylint: disable = invalid-name, W0105
+# pylint: disable = invalid-name, W0105, C0111
 """Scikit-Learn Wrapper interface for LightGBM."""
 from __future__ import absolute_import
 
 import numpy as np
-from .basic import LightGBMError, Predictor, Dataset, Booster, is_str
+from .basic import LightGBMError, is_str
```
Comment (guolinke) [inline python-package/lightgbm/sklearn.py]: Doesn't Booster?

File: python-package/lightgbm/engine.py [hunk 2]
```diff
@@ -131,12 +131,16 @@ def train(params, train_data, num_boost_round=100,
         predictor = init_model
     else:
         predictor = None
+    init_iteration = predictor.init_iteration if predictor else 0
```
Comment (guolinke) [inline python-package/lightgbm/engine.py]: I think init_iteration should be inside of Booster. Don't expose it. 
Early_stopping still return the relative iteration. 
And use init_iteration+passed_iter as final iteration to predict.

File: python-package/lightgbm/basic.py [hunk 3]
```diff
@@ -195,6 +187,7 @@ def __init__(self, model_file=None, booster_handle=None, is_manage_handle=True):
         """
         self.handle = ctypes.c_void_p()
         self.__is_manage_handle = True
+        self.init_iteration = init_iteration
```
Comment (guolinke) [inline python-package/lightgbm/basic.py]: Actually, init_iteration will be reset by out_num_iterations. 
So not need to pass this parameter.
you can just let self.__num_total_iteration to public.

File: python-package/lightgbm/basic.py [hunk 4]
```diff
@@ -854,19 +847,21 @@ def num_feature(self):
         return ret.value
 
 class Booster(object):
-    """"A Booster of of LightGBM.
+    """"A Booster of LightGBM.
     """
-    def __init__(self, params=None, train_set=None, model_file=None, silent=False):
+    def __init__(self, params=None, train_set=None, model_file=None, init_iteration=0, silent=False):
```
Comment (guolinke) [inline python-package/lightgbm/basic.py]: not need to pass init_iteration. just get by train_set.predictor.num_iteration.

File: python-package/lightgbm/sklearn.py (hunk 1 above)
Comment (wxchan) [inline python-package/lightgbm/sklearn.py]: It's unused.

Comment (guolinke) [inline python-package/lightgbm/sklearn.py]: self._Booster.predict ?

File: python-package/lightgbm/engine.py (hunk 2 above)
Comment (guolinke) [inline python-package/lightgbm/engine.py]: OK, i see. 
I think maybe we should start from init_iteration in here:
https://github.com/Microsoft/LightGBM/blob/master/python-package/lightgbm/engine.py#L212

Comment (guolinke) [inline python-package/lightgbm/engine.py]: I think the usage is:
1. Can get num_iteration from Predictor.
2. Set Init_iteration=predictor.num_iteration in Booster.
3. loop iteration into [Init_iteration, Init_iteration+num_boost_round)

Comment (wxchan) [inline python-package/lightgbm/engine.py]: I fixed it. Seem init_iteration is not needed in Booster, I already test on init_model=Booster, model_file and Predictor.

File: python-package/lightgbm/sklearn.py (hunk 1 above)
Comment (wxchan) [inline python-package/lightgbm/sklearn.py]: self._Booster comes from train(). You can try test_sklearn.py. Pylint helps detect such warnings.

**Pull request merged**